lib/internal/Magento/Framework/Module/Dir/Resolver.php

```diff
@@ -0,0 +1,49 @@
+<?php
+/**
+ * Copyright © 2015 Magento. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Framework\Module\Dir;
+
+/**
+ * Provides ability to statically register modules which do not reside in the modules directory
+ *
+ * @author Josh Di Fabio <joshdifabio@gmail.com>
+ */
+final class Resolver implements ResolverInterface
+{
+    /**
+     * Paths to modules
+     *
+     * @var string[]
+     */
+    private static $modulePaths = [];
+
+    /**
+     * Sets the location of a module. Necessary for modules which do not reside in modules directory
+     *
+     * @param string $moduleName Fully-qualified module name
+     * @param string $path Absolute file path to the module
+     */
+    public static function setModulePath($moduleName, $path)

```

Comment:
Static calls are hard-coded dependencies, they can not be substituted or decorated. So please avoid them. Other problem is that with such approach there is no strict rule when all module paths are available, because you don't know if all modules called setModulePaths, and module dev doesn't know when it should call setModulePath.